### PR TITLE
[NR] 105 Adds dependencies to the healthcheck

### PR DESCRIPTION
**Pull request by retroredge**

### CODE REVIEW

File: app/controllers/UtilityController.java [hunk 1]
```diff
@@ -3,38 +3,80 @@
...
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Stream;
+
+import static java.lang.Runtime.getRuntime;
 
 public class UtilityController extends Controller {
 
     private final String version;
 
+    private PdfGenerator pdfGenerator;
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: make final

File: app/controllers/UtilityController.java [hunk 2]
```diff
@@ -3,38 +3,80 @@
...
+    private DocumentStore documentStore;
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: make final

File: app/services/AlfrescoStore.java [hunk 3]
```diff
@@ -111,6 +117,18 @@ public AlfrescoStore(Config configuration, WSClient wsClient) {
                 });
     }
 
+    @Override
+    public CompletionStage<Boolean> isHealthy() {
+        return wsClient.url(alfrescoUrl + "noms-spg/")
```
Comment (magic-hat-solutions) [inline app/services/AlfrescoStore.java]: Use makeRequest() below for DRY code

File: app/controllers/UtilityController.java [hunk 4]
```diff
@@ -3,38 +3,80 @@
...
     }
 
-    public Result healthcheck() {
+    public CompletionStage<Result> healthcheck() {
+        val pdfGeneratorHealthFuture = pdfGenerator.isHealthy().toCompletableFuture();
+        val documentGeneratorHealthFuture = documentStore.isHealthy().toCompletableFuture();
+        val allHealthFutures = CompletableFuture.allOf(pdfGeneratorHealthFuture, documentGeneratorHealthFuture);
 
+        return allHealthFutures
+            .thenApply((ignored) -> buildResult(pdfGeneratorHealthFuture.join(), documentGeneratorHealthFuture.join()));
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: Don't need brackets round (ignored)

File: app/controllers/UtilityController.java [hunk 5]
```diff
@@ -3,38 +3,80 @@
...
+                {
+                    put("pdf-generator", pdfGeneratorStatus ? "OK" : "FAILED");
+                    put("document-store", documentStoreStatus ? "OK" : "FAILED");
+                }
+            });
+        }
+    });
+    }
+
+    private ImmutableMap<String, ? extends Number> runtimeInfo() {
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: getRuntimeInfo() for bean compliance?

File: app/controllers/UtilityController.java [hunk 6]
```diff
@@ -3,38 +3,80 @@
...
+    private Result buildResult(Boolean pdfGeneratorStatus, Boolean documentStoreStatus) {
+        return JsonHelper.okJson(new HashMap<String, Object>() {
+        {
+            put("status", pdfGeneratorStatus && documentStoreStatus ? "OK" : "FAILED");
+            put("dateTime", DateTime.now().toString());
+            put("version", version);
+            put("runtime", runtimeInfo());
+            put("fileSystems", fileSystemDetails());
+            put("localHost", localhost());
+            put("dependencies", new HashMap<String, Object>() {
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: Could use  ImmutableMap.of() here

File: app/controllers/UtilityController.java [hunk 7]
```diff
@@ -44,23 +86,15 @@ public Result healthcheck() {
...
-                        "processors", runtime.availableProcessors(),
-                        "freeMemory", runtime.freeMemory(),
-                        "totalMemory", runtime.totalMemory(),
-                        "maxMemory", runtime.maxMemory()
-                ));
-                put("fileSystems", roots);
-                put("localHost", finalLocalHost);
-            }
-        });
+    private Stream<Object> fileSystemDetails() {
```
Comment (magic-hat-solutions) [inline app/controllers/UtilityController.java]: getFileSystemDetails() ?

File: app/controllers/UtilityController.java (hunk 6 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: It only accepts up to 4 entries. Could Use ImmutableMap's builder pattern maybe

File: app/controllers/UtilityController.java (hunk 5 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: I stopped using 'get' in front of methods like this a few years ago to improve readability in the client code. I prefer things like `render.runtimeInfo()` over thing like `render.getRuntimeInfo()'. Try reading each out loud to get my point.

With some frameworks you're forced to use the 'get' 'set' pattern of course.

File: app/controllers/UtilityController.java (hunk 7 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: Same reasoning as runtimeInfo

File: app/services/AlfrescoStore.java (hunk 3 above)
Comment (retroredge) [reply app/services/AlfrescoStore.java]: Did look at that but makeRequest uses the `onBehalfOfUser` which we don't have in the context of a health endpoint been called

File: app/controllers/UtilityController.java (hunk 1 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: Done

File: app/controllers/UtilityController.java (hunk 2 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: Done

File: app/controllers/UtilityController.java (hunk 4 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: Done

File: app/controllers/UtilityController.java (hunk 6 above)
Comment (retroredge) [reply app/controllers/UtilityController.java]: Converted to the ImmutableMap builder pattern which looks nicer

**Pull request merged**